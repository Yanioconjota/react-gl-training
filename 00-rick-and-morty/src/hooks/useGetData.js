import { useState, useEffect } from "react";
import { getData } from "../helpers/getData";

export const useGetData = () => {
  
  const [state, setState] = useState({
    characters: [],
    loading: true
  });

  useEffect(() => {
    getData()
      .then(characters => {
        setState({
          characters: characters,
          loading: false,
        });
      })
  }, []);
  
  return state;
};
