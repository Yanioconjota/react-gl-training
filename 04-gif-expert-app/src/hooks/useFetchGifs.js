import { useState, useEffect } from "react";
import { getGifs } from "../helpers/getGifs";


export const useFetchGifs = (category, limit, offset) => {

  const [state, setState] = useState({
    data: [],
    loading: true,
    limit: limit,
    offset: offset,
  });

  useEffect(() => {
    getGifs(category, limit, offset)
      .then( imgs => {
        setState({
          data: imgs,
          loading: false,
          limit: limit,
          offset: offset 
        });
      });
  }, [category, limit, offset]);

  return state; //{ data: [], loading: true }
}