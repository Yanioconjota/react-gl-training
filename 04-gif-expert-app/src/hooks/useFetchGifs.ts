import { useState, useEffect } from "react";
import { getGifs } from "../helpers/getGifs";
import { IGifState } from "../modules/exports";


export const useFetchGifs = (category: string, limit: number, offset: number) => {

  const [state, setState] = useState<IGifState>({
    data: [],
    loading: true,
    limit: limit,
    offset: offset,
  });

  useEffect(() => {
    getGifs(category, limit, offset)
      .then( imgs => {
        setState({
          data: imgs,
          loading: false,
          limit: limit,
          offset: offset 
        });
      });
  }, [category, limit, offset]);

  return state; //{ data: [], loading: true }
}