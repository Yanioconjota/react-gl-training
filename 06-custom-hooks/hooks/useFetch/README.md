# useFetch Hook

### Ejemplo de uso:

```
  const url = 'url de una API'
  const { data: null, loading: true, error: null } = useFetch(url)
```

useFetch() //Devuelve un nuevo estado con la data.