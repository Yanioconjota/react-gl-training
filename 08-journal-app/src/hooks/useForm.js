import { useState } from "react";

export const useForm = (initialState = {}) => {
  //Se pueden agregar las validaciones del formulario acÃ¡
  const [values, setValues] = useState(initialState);

  const handleInputChange = ({ target }) => {
    setValues({
      ...values,
      [target.name]: target.value,
    });
  };

  const reset = () => {
    setValues(initialState);
  };

  return [values, handleInputChange, reset];
};
