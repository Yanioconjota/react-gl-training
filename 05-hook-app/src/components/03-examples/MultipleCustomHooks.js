import React from 'react';
import { Button } from 'react-bootstrap';
import { useCounter } from '../../hooks/useCounter';
import { useFetch } from '../../hooks/useFetch';

const MultipleCustomHooks = () => {

  const { counter, increment } = useCounter(1);

  const { loading, data } = useFetch(
    `https://www.breakingbadapi.com/api/quotes/${counter}`
  );

  const { author, quote } = !!data && data[0];
  
  return (
    <>
      <h1>Breaking bad Quotes</h1>
      <hr />
      {loading ? (
        <div className="alert alert-primary text-center">loading...</div>
      ) : (
        <>
          <blockquote className="blockquote Text-right">
            <p>{quote}</p>
            <footer className="blockquote-footer">{author}</footer>
          </blockquote>
          <Button onClick={ () => increment(1) } variant="primary" className="w-100">
           Cargar otra cita
          </Button>
        </>
      )}
    </>
  );
};

export default MultipleCustomHooks;
