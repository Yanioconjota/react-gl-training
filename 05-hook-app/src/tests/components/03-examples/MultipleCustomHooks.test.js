import "@testing-library/jest-dom";
import { shallow } from "enzyme";

import MultipleCustomHooks from "../../../components/03-examples/MultipleCustomHooks";
import { useCounter } from "../../../hooks/useCounter";
import { useFetch } from "../../../hooks/useFetch";

//como el useFetch ya fue testeado mockeamos su implementción para esperar que retorne información ficticia
jest.mock('../../../hooks/useFetch');

jest.mock("../../../hooks/useCounter");

describe('debe mostrar el componente correctamente', () => {
  
  beforeEach(() => {
    useCounter.mockReturnValue({
      counter: 1,
      increment: () => {},
    });
  });
  
  test('debe traer la info por defecto', () => {
    //debe mockearse la llamada al useFecth para que no rompa por el jest.mock. En este caso nos interesa que devuelva la info por defecto porque es cuando se arma el componente
    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null
    });

    const wrapper = shallow(<MultipleCustomHooks />);
    expect(wrapper).toMatchSnapshot();
  });

  test('debe mostrar la info', () => {
    useFetch.mockReturnValue({
      data: [{
        author: 'Homero Thompson',
        quote: 'Creo que le habla a usted'
      }],
      loading: false,
      error: null,
    });
    const wrapper = shallow(<MultipleCustomHooks />);
    console.log(wrapper.html());
    //Buscamos que no exista ningún elemento con la clase alert ya que solo se muestra cuando no carga la info
    expect(wrapper.find('.alert').exists()).toBe(false);
    expect(wrapper.find('p').text().trim()).toBe('Creo que le habla a usted');
    expect(wrapper.find('footer').text().trim()).toBe('Homero Thompson');
  });
  
  
});
