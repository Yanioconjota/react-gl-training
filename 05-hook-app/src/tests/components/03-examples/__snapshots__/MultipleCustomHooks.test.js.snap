// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`debe mostrar el componente correctamente debe traer la info por defecto 1`] = `
<Fragment>
  <h1>
    Breaking bad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-primary text-center"
  >
    loading...
  </div>
</Fragment>
`;

exports[`debe mostrar el componente correctamente should first 1`] = `
<Fragment>
  <h1>
    Breaking bad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-primary text-center"
  >
    loading...
  </div>
</Fragment>
`;
